### ISSUE TITLE
Allow marking specific layers to be treated as full "top layers"

### ISSUE THREAD

**Issue by stew675** [SoftFever/OrcaSlicer#5738]
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

When printing objects that have large flat areas with additional surface features on top, the slicer will try to extend the infill vertically into the top features.  This results in the layer that the features sit on top of having an inconsistent appearance as the slicer has to leave gaps in the surface for the lower infill to extend vertically into the upper features.

Embossed text would be one such example where this occurs, where the layer that the text sits atop of is full of moves to exclude the areas that the embossing sits on because it wants to extend the infill vertically into the embossing.  The action of filling in the broken up surface often results in a messy appearance.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

Ideally, much like how we can right-click on a layer height to add in custom g-code, there would be an option to say "Treat as a top layer".

Basically the slicer would treat the object up to that layer as if it were a complete object in its own right, and additional features on top of that as before.

The best way to describe this would work would be like the following gcode splicing example:

- Slice an object normally.
- Now cut the object at the layer that you want to have a clean appearance
- Reslice the object up to the cut as if it were a complete object with its top layer at that height.
- Remove the PRINT_END portion of the lower-cut's gcode
- Now insert the original g-code from above that cut point onto the end of the layer-cut gcode

Now we have generated g-code that has a fully clean middle-layer that has features added atop of it, without the broken surface moves to avoid the infill that used to extend between the upper and lower portions.

This will result in much cleaner prints without having to do the above g-code splicing work by hand.

**Comment by jeremytodd1** [comment]
I think I know what you're asking for and I believe the feature you _actually_ want is SuperSlicer's "**Anchor Solid Infill by X mm**" setting

This specific feature is a better solution to this issue, and is the one feature that I have desperately wanted for Orca to get for about a year now. This single setting has huge impacts on top layers for some models, especially for things like signs with text on the top, or even maps that people have been printing lately.

Here is a bit of an example to illustrate what "Anchor Solid Infill by X mm" does.
This is a of "The Office."
![theofficesign](https://github.com/SoftFever/OrcaSlicer/assets/6227771/a3d712df-5eba-489e-a4d7-a06111f3e7ad)

After tuning that Anchor Solid Infill by X mm setting, this is how SuperSlicer prints the top layer before the top text:
![superslicer_gif](https://github.com/SoftFever/OrcaSlicer/assets/6227771/57f860ca-cd4f-47d6-810f-d6513f4eeafb)

Orca does not have this feature at all, so it's impossible to tune for this. This is how Orca prints the same layer.
![orca_gif](https://github.com/SoftFever/OrcaSlicer/assets/6227771/2addfb65-c5b8-49b7-a91e-60f6ccf2381a)

SuperSlicer is able to print the entire layer in one go, which makes it look much more consistent.
Orca has to stop the line and continue it again, which can cause inconsistencies in the look of the layer.

@SoftFever , @igiannakas ,
TLDR: Having this Anchor Solid Infill by X mm feature would be a great way to keep the top layers looking consistent all across the board.

**Comment by Ataraxia-Mechanica** [comment]
You can add a height range modifier to that layer and the layer below it, set it to 0 top layers and 100% infill, disable only one wall on top, then slow solid infill down so its speed matches top layer. Very hacky but it works.
![image](https://github.com/SoftFever/OrcaSlicer/assets/111327195/9dd3037f-75a4-4474-87a7-002932a65d61)
![image](https://github.com/SoftFever/OrcaSlicer/assets/111327195/6633653b-7e7b-4917-b26e-6e3fa5073e3a)

**Comment by jeremytodd1** [comment]
> You can add a height range modifier to that layer and the layer below it, set it to 0 top layers and 100% infill, disable only one wall on top, then slow solid infill down so its speed matches top layer. Very hacky but it works. 

Thank you for this. This is the only workaround that I've seen that works. (Just to point out, you can replicate the "only one wall on top layer" setting by just simply changing wall count to 1 in the modifier as well, so you don't need to lose that setting for the top layer)

Saying that, I think this only shows how badly Orca needs the **Anchor Solid Infill by X mm** setting. That is a lot of steps instead of just tuning one simple value.

**Comment by stew675** [comment]
> I think I know what you're asking for and I believe the feature you _actually_ want is SuperSlicer's "**Anchor Solid Infill by X mm**" setting

Thank you for that.  The videos you provided are pretty much exactly that.  The Orca Slicer video perfectly explains what Orca Slicer does today, and it's really annoying.

I'm not particularly fussed by what name it would go under, but I think it's important that the layer gets treated as a top layer, rather than as Solid Infill.  Perhaps SuperSlicer is treating it as a top layer though?

> Thank you for this. This is the only workaround that I've seen that works. (Just to point out, you can replicate the "only one wall on top layer" setting by just simply changing wall count to 1 in the modifier as well, so you don't need to lose that setting for the top layer)

Agree.  Thank you for the work around @Ataraxia-Mechanica but that is way more complicated.

> Saying that, I think this only shows how badly Orca needs the Anchor Solid Infill by X mm setting. That is a lot of steps instead of just tuning one simple value.

I'd even be happy with a radio button that simply says "Treat layers that have partial top-layer infill as a complete/full top layer", so that way it catches all layers that have top-surface features without the user having to manually fiddle with it.  The subsequent (higher) layers should start off as if there were 0 bottom layers (ie. immediately start printing infill in the next layer just like what happens currently).

**Comment by jeremytodd1** [comment]
@Ataraxia-Mechanica 

Another issue with that solution is that with the modifier, it's no longer considered a top surface.
![zhopsetting](https://github.com/SoftFever/OrcaSlicer/assets/6227771/fd51a883-89b7-4c25-a92c-555d751de1ea)

Due to this, Lift Z Enforcement's "Top and Bottom" surfaces setting will not work on this. This will cause scratches on the top layer. You can see here there is no lift in the travels:
![nozhop](https://github.com/SoftFever/OrcaSlicer/assets/6227771/354b872f-1b33-44c1-a858-0886af5da387)

**Comment by vovodroid** [comment]
@jeremytodd1 , what is exact option name for **Anchor Solid Infill by X mm**? I don't see it in SS 2.5.59

**Comment by jeremytodd1** [comment]
@vovodroid 

It looks like the actual value's name is "external_infill_margin"
<img width="412" alt="Screenshot 2024-07-26 123505" src="https://github.com/user-attachments/assets/1ab0aafb-1fec-42ae-b56c-b4142e9c91c5">

You can find it in the Infill section in the Print Settings tab.

**Comment by Eldenroot** [comment]
I believe @vovodroid will send a new PR for this soon :)

**Comment by jeremytodd1** [comment]
That would be absolutely amazing. Not being able to tune for this has been haunting me ever since I switched to Orca. I really appreciate it.

**Comment by vovodroid** [comment]
It doesn't work now in SS 2.7 branch, as function where it was implemented was changed since Prusa 2.5. Let's hope Merill will fix it eventually, thus it could be ported to Orca.

**Comment by Anollya** [comment]
> That would be absolutely amazing. Not being able to tune for this has been haunting me ever since I switched to Orca. I really appreciate it.

I fully agree!
any news here?

old:
![signal-2024-09-24-192731_002](https://github.com/user-attachments/assets/4f36bb2c-b707-4491-85f3-a2e5a54adb66)

workaround - perfect:
![signal-2024-09-24-192743_002](https://github.com/user-attachments/assets/54e97484-7f65-47df-8e9f-e7d15f687c38)

**Comment by vovodroid** [comment]
Try to set Ensure Vertical Shell Thickness to All, see #7386

**Comment by jeremytodd1** [comment]
> Try to set Ensure Vertical Shell Thickness to All, see #7386

It doesn't seem to do a whole lot for me.
My latest test file is small map I made just for testing for this issue: 
https://www.printables.com/model/1084726-test-map-for-top-layer-issue

The goal of the test model is to have layer 6 (at 1.2mm height) print in just one clean layer that doesn't have any additional details on it. Layer 7 should be the layer where buildings and roads start printing.
Here is layer 6 for reference:
![Screenshot 2024-11-23 180438](https://github.com/user-attachments/assets/6b006d9a-30a2-4d52-97cd-e2ffa4514556)

I see Prusa released an alpha version of PrusaSlicer 2.9. Do you know if this version would help improve things?

**Comment by Nilssonfpv** [comment]
Still no fix for this short of the workaround that means you have no top players at all?
We could really do with a way to mark what the "top" layer is that text/logos will be printed ontop

Id be fine if it means id just have to have the model as assembled parts, where you mark them as  "top" or "base" parts

**Comment by CodyPrince** [comment]
Any new on this? Would really like to see it in orca.

**Comment by vovodroid** [comment]
@jeremytodd1 it seems to work in latest Superslicer 2.7.61.

![Image](https://github.com/user-attachments/assets/03c8ea67-4acb-4a81-959e-9636eb330d80)

![Image](https://github.com/user-attachments/assets/55350166-346e-48e0-9079-11412f127268)

![Image](https://github.com/user-attachments/assets/cac206c9-dda3-4432-8840-f7c98bab865a)

![Image](https://github.com/user-attachments/assets/77d829c6-5141-477e-ac38-18dc5518fdd6)

Could you check this? If it was fixed there, I'll take a look on SS code and probably try port in Orca.

**Comment by jeremytodd1** [comment]
@vovodroid 

I actually helped with testing for this specific issue with the SuperSlicer devs to help get this fixed in their latest release.
Yes, I believe it has been fixed in this latest version of SuperSlicer. I've tested it with numerous models at this point (technically on an older pre-release version but it it should act the same on this new release version) and they all printed with a nice clean main layer.
Layer 11:
![Image](https://github.com/user-attachments/assets/f02b77da-af73-44b7-b89c-9c23db7c5bdd)
Layer 10:
![Image](https://github.com/user-attachments/assets/6c7bbc8b-72ee-45be-ba28-bf0de932f8ae)

I've found I need two specific settings to be set correctly for it to work.

Ensure Vertical Shell Thickness needs to be set to "Enabled (2.5)". It will not work with the other settings.
![Image](https://github.com/user-attachments/assets/f9333f1f-a4c4-4c28-b0ff-02b72e0f2c1b)

Anchor solid infill by X mm is still necessary.
![Image](https://github.com/user-attachments/assets/d39af3e4-58c3-4f2d-a0cf-ccb6f64bc22b)

I typically leave the anchor solid setting to a value of "20" (not a percentage, just a hard value of 20) and it seems to work for most models. This value can differ though. For this office sign for example, I can set that value as low as 6 and the main layer will still print clean. If I set it lower, like to even at 5, it'll start showing artifacts in the main layer.

I would be thrilled if this was now able to be ported to OrcaSlicer. This issue still **very** real to me in Orcaslicer and would love for it to be fixed.

**Comment by jeremytodd1** [comment]
@vovodroid Just curious, did you happen to look into this yet?

**Comment by vovodroid** [comment]
Still no, but certainly will.

**Comment by vovodroid** [comment]
Well, it's turn out to be more complicated than I guessed, as SS code base is different from Orca. I built ported code, but it broke functionality.

Here is the branch, may be somebody could help https://github.com/vovodroid/OrcaSlicer/tree/infill-margin-wip

@igiannakas ?

**Comment by MrMonkey1302** [comment]
@vovodroid Hey, did anyone end up helping with this at all?

**Comment by metinkale38** [comment]
I found a very simple workaround for the Emboss Tool in Orca:
 
You can set the property "From Surface" to 0.02 (lower values will be rounded to 0) and it will print the top layer without interruptions.

**Comment by jeremytodd1** [comment]
Everyone, We finally have progress on this issue but they need more people to take a look. 
Check out this PR: https://github.com/OrcaSlicer/OrcaSlicer/pull/12187

**Comment by ianalexis** [comment]
In current nigthly we have surface expansion https://www.orcaslicer.com/wiki/print_settings/strength/strength_settings_top_bottom_shells.html#surface-expansion
Is this enough or do you also prefer a mm2 based remover?

**Comment by igiannakas** [comment]
IMHO mm expansion is sufficient. Mm2 remover is less relevant as it varies the features removed based on surface area covered which is counter to the point that small features need to be merged.

If im making any sense :)

**Comment by ianalexis** [comment]
> IMHO mm expansion is sufficient. Mm2 remover is less relevant as it varies the features removed based on surface area covered which is counter to the point that small features need to be merged.
> 
> If im making any sense :)

Same, for that reasong and to be simmilar to other slicers (but this is a little bit better implementation) i did that.
But maybe in some cases instead of expanding all, a specific Threshold as the one for "Only one wall on top" may be useful.
